pkg/registry/core/service/storage/storage.go

```diff
@@ -163,12 +163,22 @@ func (r *GenericREST) defaultServiceList(serviceList *api.ServiceList) error {
 	return nil
 }
 
-// defaults a single service
-func (r *GenericREST) defaultAServiceOnRead(service *api.Service) error {
+// defaultOnReadService defaults a single Service.
+func (r *GenericREST) defaultOnReadService(service *api.Service) error {
 	if service == nil {
 		return nil
 	}
 
+	// We might find Services that were written before ClusterIP became plural.
+	// We still want to present a consistent view of them.
+	// NOTE: the args are (old, new)
+	svcreg.NormalizeClusterIPs(nil, service)
+
+	// The rest of this does not apply unless dual-stack is enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
```

Comment:
gate should be on `defaultOnRead(..)` first thing. Instead of for-each-item-on-list.

Reply:
The question is whether we expect `clusterIPs` to have a value when the gate is disabled?  Based on the logic in `dropServiceDisabledFields()`, I assumed that was your intention:

```
if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && !strategy.serviceDualStackFieldsInUse(oldSvc) {
        newSvc.Spec.IPFamilies = nil
        newSvc.Spec.IPFamilyPolicy = nil
        if len(newSvc.Spec.ClusterIPs) > 1 {
            newSvc.Spec.ClusterIPs = newSvc.Spec.ClusterIPs[0:1]
        }
    }
```

If we leave that as-is, then we really need to populate it on read (regardless of the gate).  If we don't populate on read (when gated off), we should just set it to nil when writing.  It either always has a value or never.  But lots of internal logic depends on it having a value, so this seemed cleaner.

Reply:
oh yes. I missed that one.